src/psij/executors/flux.py

```diff
@@ -50,175 +57,110 @@ def __init__(self, url: Optional[str] = None,
-            jspec.arguments = []
-
-        if not isinstance(jspec.resources, ResourceSpecV1):
-            raise InvalidJobException("resource is not ResourceSpecV1")
-        rspec = jspec.resources
-        attr = jspec.attributes
-
-        flux_spec = {
-            "tasks": [{
-                "slot": "task",
-                "count": {
-                    "per_slot": 1
-                },
-                "command": [jspec.executable] + jspec.arguments
-            }],
-            "attributes": {
-                "system": {
-                    # FIXME: flux hangs on `None` values
-                    # "cwd": jspec.directory,
-                    "duration": attr.duration.total_seconds()
-                }
-            },
-            "version": 1,
-            "resources": [{
-                "count": rspec.process_count,
-                "type": "slot",
-                "label": "task",
-                "with": [{
-                    "count": rspec.cpu_cores_per_process,
-                    "type": "core"
-                }, {
-                    "count": rspec.gpu_cores_per_process,
-                    "type": "gpu"
-                }]
-            }]
-        }
-
-        return flux_spec
+        assert job.spec
+        assert job.spec.attributes
+        if isinstance(job.spec.resources, ResourceSpecV1):
+            resources = job.spec.resources
+        elif isinstance(job.spec.resources, ResourceSpec):
+            raise InvalidJobException(
+                f"ResourceSpec version {job.spec.resources.version} not supported"
+            )
+        else:
+            resources = ResourceSpecV1(process_count=1, cpu_cores_per_process=1)
+        if resources.exclusive_node_use:
+            warnings.warn(
+                "Flux does not support exclusive_node_use=True, ignoring...",
+                UserWarning,
+            )
+        if resources.processes_per_node:
+            raise InvalidJobException("Flux does not support processes_per_node")
+        if not job.spec.executable:
+            raise InvalidJobException("Job must have an executable")
+        argv = list(job.spec.arguments) if job.spec.arguments else []
+        argv.insert(0, job.spec.executable)
+        flux_jobspec = flux.job.JobspecV1.from_command(
+            argv,
+            num_tasks=resources.process_count,
+            cores_per_task=resources.cpu_cores_per_process,
+            gpus_per_task=resources.gpu_cores_per_process,
+            num_nodes=resources.node_count,
+        )
+        if job.spec.stdout_path:
+            flux_jobspec.stdout = job.spec.stdout_path
+        if job.spec.stdin_path:
+            flux_jobspec.stdin = job.spec.stdin_path
+        if job.spec.stderr_path:
+            flux.jobspec.stderr = job.spec.stderr_path
+        flux_jobspec.duration = job.spec.attributes.duration.total_seconds()
+        fut = self._flux_executor.submit(flux_jobspec)
+        self._add_flux_callbacks(job, fut)
```

Comment:
That looks like a bit of a race condition.

Reply:
It isn't. The executor stores all the events it has seen and if you add a callback after the executor has handled it, the callback will be invoked immediately. Otherwise we (and concurrent.futures which operates the same way) would be screwed, because you can't get a handle to a future until you've submitted a job.

Reply:
- [x] &nbsp;